### ISSUE TITLE
v2.6.7: ne.evaluate() deletes global variables

### ISSUE THREAD

**Issue by crackwitz** [pydata/numexpr#313]
```
>>> foo = 10
>>> import numexpr
>>> numexpr.evaluate("2+2")
array(4, dtype=int32)
>>> foo
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'foo' is not defined
```

did not happen in v2.6.6. can you reproduce this?

**Comment by jcmgray** [comment]
I too am having this issue, or a variant of, only with 2.6.7.

```python
>>> import numexpr as ne
>>> a = ne.evaluate("1+2")
>>> ne
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'ne' is not defined
>>> exit()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'exit' is not defined
```
Tested using conda, python 3.6 on linux (both ubuntu and CentOS).

**Comment by crackwitz** [comment]
diff between commits: https://github.com/pydata/numexpr/compare/e7411f8e3fd8eb518dc5eaccf990591ed5df849e...f556c1f8a1c248f357042110795fac493a153825

**Comment by robbmcleod** [comment]
Can you try building https://github.com/pydata/numexpr/commit/b8f0d8dfee29f82d3cd7b52782be086a1c900337 and see if that fixes the bug? It appears, to me anyway, that if you call `.f_locals` on the top-most frame, you actually get the `globals()` dict back. So the memory leak that we were trying to deal with in #310 doesn't seem to appear in that case and clearing the globals dict is rather bad.

This was my test code:

    import sys
    import numexpr as ne

    def test_context():
        a = 10
        print(a, sys.getrefcount(a))
        ne.evaluate('1+1')
        print(a, sys.getrefcount(a))
        func_frame = sys._getframe(0)
        print('Function frame back:', func_frame.f_back)
        top_frame = sys._getframe(1)
        print('Top frame back:', top_frame.f_back)

    test_context()

    b = 99
    print('b', sys.getrefcount(b))
    ne.evaluate('2+2')
    print('b', sys.getrefcount(b))

**Comment by crackwitz** [comment]
b8f0d8d appears to leave globals alone. the output of your snippet looks sensible too.

```
10 73
10 73
Function frame back: <frame object at 0x00000000122E51E8>
Top frame back: None
b 7
array(4, dtype=int32)
b 7
```

**Comment by lgarrison** [comment]
Would another way to check if the locals dict is actually the global dict be to compare `f_locals` to `f_globals`?  If they're the same dict, then we don't want to clear.  For example:

```python
import sys
frame = sys._getframe(0)
fl = frame.f_locals
fg = frame.f_globals

print('f_locals is f_globals outside of function: ', fl is fg)

def test_func():
    frame = sys._getframe(0)
    fl = frame.f_locals
    fg = frame.f_globals
    
    print('f_locals is f_globals inside function: ', fl is fg)
    
test_func()
```
```
f_locals is f_globals outside of function:  True
f_locals is f_globals inside function:  False
```

Maybe it's equivalent to `call_frame.f_back is not None` in most (all?) cases, but the case where locals and globals are the same is what we really want to watch out for.

**Comment by robbmcleod** [comment]
@lgarrison based on my limited understanding of how Python stacks frames, and the documentation that is available, only the top-frame can have `.f_back == None`.

**Comment by robbmcleod** [comment]
I can prepare an immediate bug-fix release, but I would really appreciate it if the participants in this thread could test the new commits in production code. I added a test, but `unittest` is going to be quite iffy in its ability to deal with scope and frames due to how it works.

**Comment by lgarrison** [comment]
In a Jupyter notebook, the "cell" frame is not the top frame, but the locals are indeed equal to the globals.  So the `f_back` detection would not work in that case.

One can see this by running this in a Jupyter notebook:
```python
import sys
frame = sys._getframe(0)
print(frame.f_back)
print(frame.f_locals is frame.f_globals)
```
```
<frame object at 0x563212174b68>
True
```

On the command line, the output is
```
None
True
```

**Comment by jcmgray** [comment]
Yes, with the current master (f0ab31fdf03594884dcdc002f5aa7eda3b67ad13), the issue remains for ``ipython`` and ``jupyter-notebook`` cells but is fixed for the standard python interpreter.

**Comment by robbmcleod** [comment]
I made a change with 742628e to do a comparison to `f_global`.  I can say my regression tests I added to the `unittest` suite do _not_ work as intended.  Here's my test of the permutations:

    import sys
    import numexpr as ne
    def test_context():
        a = 10
        print('1: a =', a, 'refs=', sys.getrefcount(a))
        ne.evaluate('1+1')
        print('2: a =', a, 'refs=', sys.getrefcount(a))
    test_context()

    b = 99
    print('1: b =', b, 'refs=', sys.getrefcount(b))
    ne.evaluate('2')
    print('2: b =', b, 'refs=', sys.getrefcount(b))
    ne.evaluate('3', local_dict={})
    print('3: b =', b, 'refs=', sys.getrefcount(b))
    ne.evaluate('4', global_dict={})
    print('4: b =', b, 'refs=', sys.getrefcount(b))
    ne.evaluate('5', local_dict={}, global_dict={})
    print('5: b =', b, 'refs=', sys.getrefcount(b))

If anyone has thoughts on how to get `unittest` to work in the top-most frame let me know.

**Comment by robbmcleod** [comment]
Oh it's no biggy, it's not like it was intentional. I myself reviewed the Python docs for frame objects before I published the version and I didn't see anything that gave me pause.

I do sort of wonder if this has some performance issues if for Python if you're running some data mashing script from the top frame and thereby are using the globals dict a lot, instead of the fast locals.

I've got a test implemented now so if it passes CI then hopefully I can get Travis building the wheels this evening.
